admin/server/bookmarks.go

```diff
@@ -0,0 +1,127 @@
+		return &adminv1.ListBookmarksResponse{}, nil
+	}
+
+	// Error if authenticated as anything other than a user
+	if claims.OwnerType() != auth.OwnerTypeUser {
+		return nil, fmt.Errorf("not authenticated as a user")
+	}
+
+	bookmarks, err := s.admin.DB.FindBookmarks(ctx, req.ProjectId, req.UserId)
+	if err != nil {
+		return nil, status.Error(codes.Internal, err.Error())
+	}
+
+	dtos := make([]*adminv1.DashboardBookmark, len(bookmarks))
+	for i, bookmark := range bookmarks {
+		dtos[i] = bookmarkToPB(bookmark)
+	}
+
+	return &adminv1.ListBookmarksResponse{
+		DashboardBookmark: dtos,
+	}, nil
+}
+
+// GetBookmark server returns the bookmark for the user per project
+func (s *Server) GetBookmark(ctx context.Context, req *adminv1.GetBookmarkRequest) (*adminv1.GetBookmarkResponse, error) {
+	// Return an empty result if not authenticated.
+	claims := auth.GetClaims(ctx)
+	if claims.OwnerType() == auth.OwnerTypeAnon {
+		return &adminv1.GetBookmarkResponse{}, nil
+	}
+
+	// Error if authenticated as anything other than a user
+	if claims.OwnerType() != auth.OwnerTypeUser {
+		return nil, fmt.Errorf("not authenticated as a user")
+	}
+
+	bookmark, err := s.admin.DB.FindBookmark(ctx, req.BookmarkId)
+	if err != nil {
+		return nil, status.Error(codes.Internal, err.Error())
+	}
+
+	return &adminv1.GetBookmarkResponse{
+		DashboardBookmark: bookmarkToPB(bookmark),
+	}, nil
+}
+
+// CreateBookmark server creates a bookmark for the user per project
+func (s *Server) CreateBookmark(ctx context.Context, req *adminv1.CreateBookmarkRequest) (*adminv1.CreateBookmarkResponse, error) {
+	// Return an empty result if not authenticated.
+	claims := auth.GetClaims(ctx)
+	if claims.OwnerType() == auth.OwnerTypeAnon {
+		return &adminv1.CreateBookmarkResponse{}, nil
+	}
+
+	// Error if authenticated as anything other than a user
+	if claims.OwnerType() != auth.OwnerTypeUser {
+		return nil, fmt.Errorf("not authenticated as a user")
+	}
+
+	bookmark, err := s.admin.DB.InsertBookmark(ctx, &database.InsertBookmarkOptions{
+		DisplayName:   req.DisplayName,
+		Data:          req.Data,
+		DashboardName: req.DashboardName,
+		ProjectID:     req.ProjectId,
+		UserID:        req.UserId,
+	})
+	if err != nil {
+		return nil, status.Error(codes.Internal, err.Error())
+	}
+
+	return &adminv1.CreateBookmarkResponse{
+		DashboardBookmark: bookmarkToPB(bookmark),
+	}, nil
+}
+
+// RemoveBookmark server removes a bookmark for bookmark id
+func (s *Server) RemoveBookmark(ctx context.Context, req *adminv1.RemoveBookmarkRequest) (*adminv1.RemoveBookmarkResponse, error) {
+	// Return an empty result if not authenticated.
+	claims := auth.GetClaims(ctx)
+	if claims.OwnerType() == auth.OwnerTypeAnon {
+		return &adminv1.RemoveBookmarkResponse{}, nil
+	}
+
+	// Error if authenticated as anything other than a user
+	if claims.OwnerType() != auth.OwnerTypeUser {
+		return nil, fmt.Errorf("not authenticated as a user")
+	}
+
+	err := s.admin.DB.DeleteBookmark(ctx, req.BookmarkId)
```

Comment:
First get the bookmark and check that the `OwnerID` matches the bookmark's user ID